### DISCUSSION TITLE
Parameter optimisation for periodic steady-state how?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#8609] [Q&A]
I have a thermal system that exhibits periodic behaviour with some settling, and I want to find the periodic steady state (PSS) quicker than running the simulation for a long time. 
This is an illustrative toy example (source below):
![image](https://user-images.githubusercontent.com/327442/155737251-92230f02-be9f-4c0d-ab04-1acd1bac5bd2.png)

![image](https://user-images.githubusercontent.com/327442/155739645-765a20bc-2bc7-4c61-b81a-6fb5f6b06795.png)

You can see that there is a 2Hz oscillation, and after about 10s the situation has stabilised into the PSS. 
The simulation starts at some temperature `T_init`, and the PSS is characterised by the fact that the temperature at the end of a period is identical to `T_init`, so I think an optimisation routine could find the appropriate `parameter T_init`.

As OMOptim is broken and unsupported, I thought I'd use the new [annotation-based dynamic optimisation](https://www.openmodelica.org/doc/OpenModelicaUsersGuide/latest/optimization.html#built-in-dynamic-optimization-using-annotations) approach.
I realise that if I define a "Mayer" cost function for the squared difference between `T` after N periods and `T_init` (and probable another component for `der(T)`), the optimiser should find the appropriate `T_init`.

However, where I'm stumped is that said optimisation works with a time-variable input function (`u(t)`), and I can't find a way to make this accomodate the (simpler?) problem of varying only a `parameter` (like OMOptim would do). 

Does anybody have an idea how to accomplish this?

Stretch question: In the final problem I have more than one state, and to have periodic behaviour I think that all state variables and their derivatives must agree between t0 and t_f. It would be nice if there were an automated way to describe that, as that would yield a very nice capability of simulating periodic steady-state problems for OM! For now I have resigned to the fact that I probably have to enumerate all states manually and assemble an appropriate cost function.

Base model:
```modelica
model pSS_dummy_thermal2 "Periodic steady state toy problem"
  Modelica.Thermal.HeatTransfer.Sources.PrescribedHeatFlow heat_source annotation(
    Placement(transformation(extent = {{-55, -10}, {-35, 10}})));
  Modelica.Thermal.HeatTransfer.Components.HeatCapacitor heatCapacitor1(C = 5, T(displayUnit = "K")) annotation(
    Placement(transformation(extent = {{-35, 5}, {-15, 25}})));
  Modelica.Thermal.HeatTransfer.Components.ThermalResistor thermalResistor1(R = 0.5) annotation(
    Placement(transformation(extent = {{-10, -10}, {10, 10}})));
  Modelica.Thermal.HeatTransfer.Sources.FixedTemperature fixedTemperature1(T(displayUnit = "K") = 300) annotation(
    Placement(transformation(extent = {{40, -10}, {20, 10}})));
  Modelica.Blocks.Sources.Sine power(f = 1/period, offset = 1) annotation(
    Placement(visible = true, transformation(origin = {-80, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  parameter Real period = 0.5 "Oscillation period";
  parameter Real T_init = 300;  // 300.48 results in an approximately periodic solution
initial equation
  heatCapacitor1.T = T_init;
equation
  connect(heat_source.port, heatCapacitor1.port) annotation(
    Line(points = {{-35, 0}, {-30, 0}, {-25, 0}, {-25, 5}}, color = {191, 0, 0}, thickness = 0.0625));
  connect(heatCapacitor1.port, thermalResistor1.port_a) annotation(
    Line(points = {{-25, 5}, {-25, 0}, {-15, 0}, {-10, 0}}, color = {191, 0, 0}, thickness = 0.0625));
  connect(thermalResistor1.port_b, fixedTemperature1.port) annotation(
    Line(points = {{10, 0}, {15, 0}, {20, 0}}, color = {191, 0, 0}, thickness = 0.0625));
  connect(power.y, heat_source.Q_flow) annotation(
    Line(points = {{-68, 0}, {-54, 0}}, color = {0, 0, 127}));
  annotation(
    uses(Modelica(version = "4.0.0")),
    experiment(StartTime = 0, StopTime = 10, Tolerance = 1e-06, Interval = 0.02));
end pSS_dummy_thermal2;
```
Optimisation attempt
```modelica
model pss_toy2_opt
  extends pSS_dummy_thermal2;
    parameter Integer nr_periods = 2;
    Real totalCost = (heatCapacitor1.T - T_init)^2 "After nr_periods oscillations we want the temperature to be identical" annotation(isMayer = true); 
  // what to do here?
equation
  annotation(
    uses(Modelica(version = "4.0.0")),
    __OpenModelica_simulationFlags(s = "optimization", optimizerNP="1"), //optimizerNP is 1 or 3
    __OpenModelica_commandLineOptions = "+g=Optimica",
    experiment(StartTime = 0, StopTime = nr_periods*period, Tolerance = 1e-07, Interval = 0.02));
end pss_toy2_opt;
```

**Comment by lucas-mior** [comment]
@bilderbuchi  In my experience, optimization with Modelica is a lost battle (as of 2024). You should optimize using more supported tools like
a python or C numerical library (I suggest looking into scipy or GNU Scientific Library). You may even me able to call into your Modelica model by using FMU or running the model executable and checking the result file. It's cumbersome but it is flexible.

**Reply by casella** [reply]
The Optimica extension of OMC is highly experimental (and currently not actively developed), so I agree that using it is not a nice experience. However, I would not say that optimization with Modelica (not with OpenModelica!) is a lost battle. On the contrary, I believe there is a lot of potential in using Modelica tools (including OpenModelica) as a modelling frontend for tools such as CasADI, Pyomo, that give you access to more set-up, tuning and optimization features than simple drag-drop-click-to-run interface can provide.

@matteodepascali is currently working on an interface between OMC's Base Modelica output and the Pyomo.DAE tool, with the goal of using CONOPT to solve the problem. We are talking of nonlinear models with more than 5,000 equations and up to 100,000 equations. Early results on a definitely non-trivial system (see: https://ieeexplore.ieee.org/document/10252534 and https://github.com/looms-polimi/SOFCPoliMi, ) are very encouraging.

**Reply by bilderbuchi** [reply]
That sounds interesting. Will that be able differentiate into the Modelica code, to get gradients w.r.t. optimizing parameters, or will that be black-box?

**Reply by matteodepascali** [reply]
There are several possibilities employing Python:
- CasADi may take as input an FMU and get analytical Jacobians from the FMI formulation and then evaluate numerical Hessians [(Ref)](https://web.casadi.org/docs/#the-daebuilder-class);
- Currently I'm parsing the BaseModelica code generated by OpenModelica to formulate a Pyomo optimization problem. It takes advantage of the equation-based formulation obtained directly from Modelica to obtain analytical Jacobians and Hessians.

I think that the same approach I'm using for Pyomo might be used also with CasADi. This second strategy makes it easier to further customize the optimization problem once the main structure has been obtained through a parsing script.

**Comment by felix-sil3x** [comment]
Hello @bilderbuchi 
I don't think I can answer your optimization questions but, if the problems remains, I can share my experience. Maybe it will help you.
I was looking for the PSS like you, with a periodic excitation of my system. The objective was to trigger a transient once the periodic regime was reached.
I then used the Fast Fourier Transform (FFT) applied to the periodic output signals. I subsequently utilized the terminate simulation function with a stop criterion based on a sufficiently low time derivative of my amplitudes. From there, I could initialize my transient model using the average value obtained from the steady-state calculation.

**Reply by bilderbuchi** [reply]
Thank you! This seems to be a related, but not identical problem -- I want to avoid simulating until the behaviour "settles", and get the settled periodic state directly by optimizing on start parameters. 
Using `terminate` keyed on the FFT results is a nice approach for automated termination, though!
